Return vector distances of ANN queries in CQL

QuerthDP opened the pull request:
### Motivation

It should be possible to return the distance in CQL statements following the [Cassandra compatible syntax](https://cassandra.apache.org/doc/latest/cassandra/getting-started/vector-search-quickstart.html#query-vector-data-with-cql):

```
SELECT comment, similarity_cosine(comment_vector, [0.1, 0.15, 0.3, 0.12, 0.05])
    FROM cycling.comments_vs
    ORDER BY comment_vector ANN OF [0.1, 0.15, 0.3, 0.12, 0.05]
    LIMIT 1;
```

The distance is already returned in the Vector Store API. 

---

### Solution

The patch adds functions:
- `similarity_cosine(<vector_column>, <vector_value>)`,
- `similarity_euclidean(<vector_column>, <vector_value>)`,
- `similarity_dot_product(<vector_column>, <vector_value>)`

Every function matching the `similarity_function` option in `vector_index` respectively.

The arguments provided to the function are required to be the same as the values in `ORDER BY … ANN OF …` clause.

The logic of forwarding the `ann_result` from `select_statement` to `execute` in `do_evaluate` of `function_call` is implemented.
The `query_options::specific_options` are used.

---

### Tests

Integration tests for the feature: https://github.com/scylladb/vector-store/pull/241

---

Fixes: VECTOR-181

New feature, should land into 2026.1, no backport needed.

knowack1 commented on cql3/expr/expression.cc:
```diff
@@ -1640,6 +1641,24 @@ static cql3::raw_value do_evaluate(const usertype_constructor& user_val, const e
+
+        // Check if partition key matches
+        if (partition_key.size() == inputs.partition_key.size() &&
+            std::equal(partition_key.begin(), partition_key.end(), inputs.partition_key.begin())) {
```
```
if (partition_key.size() == inputs.partition_key.size() &&
            std::equal(partition_key.begin(), partition_key.end(), inputs.partition_key.begin())) {
```
->
```
if (partition_key == inputs.partition_key) {
```

QuerthDP replied:
The `inputs.partition_key` is of type `std::span<const bytes>` which does not define operator `==`.

knowack1 commented on cql3/expr/expression.cc:
```diff
@@ -1640,6 +1641,24 @@ static cql3::raw_value do_evaluate(const usertype_constructor& user_val, const e
+        if (partition_key.size() == inputs.partition_key.size() &&
+            std::equal(partition_key.begin(), partition_key.end(), inputs.partition_key.begin())) {
+            if (clustering_key.size() == inputs.clustering_key.size() &&
+                std::equal(clustering_key.begin(), clustering_key.end(), inputs.clustering_key.begin())) {
```
same as above

knowack1 commented on cql3/expr/expression.cc:
```diff
@@ -1640,6 +1641,24 @@ static cql3::raw_value do_evaluate(const usertype_constructor& user_val, const e
 
+static size_t find_matching_key_index(const evaluation_inputs& inputs, const std::vector<cql3::statements::primary_key>& keys) {
+    for (size_t i = 0; i < keys.size(); ++i) {
+        auto primary_key = keys[i];
```
Is it intended to copy primary_key here? Not sure how expensive is this copy.

knowack1 commented on service/vector_store_client.cc:
```diff
@@ -172,11 +174,20 @@ auto ck_from_json(rjson::value const& item, std::size_t idx, schema_ptr const& s
 }
 
+auto distance_from_json(rjson::value const& item, std::size_t idx) -> std::expected<float, ann_error> {
+    auto const& distance_json = item.GetArray()[idx];
```
Would be better to call item.GetArray() once, instead of every iteration.

knowack1 commented on cql3/statements/select_statement.cc:
```diff
@@ -1221,24 +1221,29 @@ indexed_table_select_statement::actually_do_execute(query_processor& qp,
+        auto [pkeys, distances] = ann_result.value();
+
+        std::unique_ptr<cql3::query_options> internal_options = std::make_unique<cql3::query_options>(cql3::query_options(options));
+        internal_options.reset(new cql3::query_options(std::move(internal_options), std::move(ann_result).value()));
```
consider something like:
```
auto internal_options = std::make_unique<cql3::query_options>(
    std::make_unique<cql3::query_options>(options), // Create a copy to be moved from
    ann_result.value()
);
```

knowack1 commented on cql3/expr/expression.cc:
```diff
@@ -1670,7 +1689,30 @@ static cql3::raw_value do_evaluate(const function_call& fun_call, const evaluati
 
-    bytes_opt result = scalar_fun->execute(arguments);
+    bytes_opt result;
+    if (dynamic_cast<functions::vector_similarity_fct*>(scalar_fun)) {
```
Consider avoiding vector_similarity_fct-specific code here in the generic implementation. This approach removes the benefit of the polymorphism design previously encouraged in this code.

As a solution, consider either to:
1 extending execute as follows:
```
scalar_fun->execute(arguments, inputs.options);
```
2. or try to pass the required data via `arguments`  (if possible).

QuerthDP replied:
I will process with the option 1. as it feels reasonable to preserve the polymorphism.
I believe option 2. is not quite possible as these are the arguments passed by the user via CQL.

knowack1 commented on cql3/functions/vector_search_fcts.hh:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    bytes_opt execute(std::span<const bytes_opt> parameters, const expr::evaluation_inputs& inputs) override {
+        SCYLLA_ASSERT(parameters.empty());
```
is this assert safe? Doesn't it depend on user input?

QuerthDP replied:
It does depend on user input and we want the user not to put any input as parameters.
But it's a good catch - we'd rather throw an exception here, not abort Scylla :)

QuerthDP commented:
Changelog:
- move change of `execute(arguments)` -> `execute(arguments, inputs)` into proper commit

knowack1 commented on cql3/functions/vector_search_fcts.cc:
```diff
@@ -0,0 +1,58 @@
+namespace cql3 {
+namespace functions {
+
+static size_t find_matching_key_index(const expr::evaluation_inputs& inputs, const std::vector<cql3::statements::primary_key>& keys) {
```
prefer anonymous namespace as this is C++

nyh replied:
static is perfectly good C++.

swasik commented:
@nyh - this one is ready for review

swasik commented:
This is a simpler version of `similarity_xxx` functions from Cassandra ([see here](https://cassandra.apache.org/doc/latest/cassandra/getting-started/vector-search-quickstart.html#query-vector-data-with-cql)). However, in this case we do not need the metric name and the vector as vector is already provided in the ANN part and metric is provided when creating the index.  So the example usage will be:

```sql
SELECT comment, vector_similarity()
    FROM cycling.comments_vs
    ORDER BY comment_vector ANN OF [0.1, 0.15, 0.3, 0.12, 0.05]
    LIMIT 3;
```

and apart from the content of `comment` column it will also return the distance between `comments_vs` and `[0.1, 0.15, 0.3, 0.12, 0.05]`. This does not have to be calculated in Scylla - it is returned from the Vector Store.

As about documentation - it is on the way - we extracted it to separate tasks to provide the holistic documentation for the whole feature. Today we merged documentation for creating indexes, tomorrow we will add ANN documentation (VECTOR-247) and next week quick start and feature documentation.

nyh commented on cql3/query_options.hh:
```diff
@@ -77,6 +78,7 @@ public:
         const std::optional<db::consistency_level> serial_consistency;
         const api::timestamp_type timestamp;
         const service::node_local_only node_local_only;
+        const vector_search::vector_store_client::ann_result ann_result;
```
It's not clear to me how ann **results** fits into **query** options. There aren't query options, we don't know them at query time - they are results...

Maybe @avikivity has a better intuition than me here about what this query_options is about.

nyh commented on cql3/statements/select_statement.cc:
```diff
@@ -1016,6 +1016,20 @@ indexed_table_select_statement::prepare(data_dictionary::database db,
     if (prepared_ann_ordering.has_value()) {
+
+        // Add partition and clustering key columns to process ANN ordering.
+        // They are needed to compare the result row primary key with ann result to return distance properly.
```
I didn't understand (from the commit message or this comment) **why** the primary key is needed again at this stage. My only guess is that perhaps we have filtering which filters out only some of the rows, and then we need to find the matching ANNs in the ANN vector? Is this guess correct?

I am getting the feeling (not backed up by any real experience) that we did things the wrong way: Instead of keeping a results list and distances list separately, we could have had a list of pairs - primary key and distance. Later when you read additional columns, you get a tuple of columns and distance - one tuple for each row. Then, when you filter, you get rid of whole tuples - the row together with its distance. When you want to read the vector_distance() that's just the last value in that tuple, just like the column "x" may be the third value in that tuple.

QuerthDP replied:
As I believe we're turning back to the Cassandra's syntax that includes the pk columns as an argument. The issue will not be present anymore.

QuerthDP replied:
> I am getting the feeling (not backed up by any real experience) that we did things the wrong way: Instead of keeping a results list and distances list separately, we could have had a list of pairs - primary key and distance.

That was the case that struggled me the whole time implementing this.
I was wondering if having a vector of pairs instead of pair of vectors was a better idea, but I believe it doesn't change things much. Of course it's more intuitive for me to have iterated through pairs of pk and distances, but I didn't see the impact on complexity taking one or another approach.
The main reason why I've chosen the pair of vectors is that it's the exact result we're receiving from the Vector Store service. Not really sure why, but wanted to fill in.

>  Later when you read additional columns, you get a tuple of columns and distance - one tuple for each row. Then, when you filter, you get rid of whole tuples - the row together with its distance. When you want to read the vector_distance() that's just the last value in that tuple, just like the column "x" may be the third value in that tuple.

I don't quite understand the idea here. Do you mean that we should have added the distance column for every ANN query, and only later on decide whether would we print the column or not depending on  if the function is called?

nyh replied:
The main difference I see between two lists (keys and distances) and one list of (key, distance) pairs is that in the latter filtering will eliminate some of these pairs easily, but in the former (which you used) if filtering deletes some of the keys, you need to know how to find the distance for each of the remaining keys - and that resulted in the O(N^2) (where N = page size) algorithm you used.
 
Ideally, if the query doesn't ask for the actual distance, Scylla would tell the vector search engine that it's not interested in the distances, and it will return just keys instead of (key,distance) pairs. But that's an optimization.

avikivity commented:
> ### Motivation
> It should be possible to return the distance in CQL statements following the simplified syntax:
> 
> ```
> SELECT comment, vector_similarity()
>     FROM cycling.comments_vs
>     ORDER BY comment_vector ANN OF [0.1, 0.15, 0.3, 0.12, 0.05]
>     LIMIT 3;
> ```
> 

The syntax `vector_similarity()` is lacking, it is not a pure function (it requires to look at context to understand it).

Better syntax is `vector_similarity(comment_vector, [...])`.

Of course, this syntax allows comparing vector similarity of unrelated stuff:

```cql
SELECT comment, vector_similarity(image_vector, :vec2)
FROM ...
ORDER BY comment_vector ANN OF :vec1
LIMIT 3
```

(here, both image_vector doesn't match comment_vector, and :vec2 doesn't match :vec1)

We can require that the parameters match.

With this syntax, we could have avoided the ANN OF operator:

```cql
SELECT comment, vector_similarity(comment_vector, :vec)
FROM ...
ORDER BY vector_similarity(comment_vector, :vec) DESC
LIMIT 3
```

But ANN OF is a given now.


To compare, you can run this query in postgres:

```
SELECT x, sin(x) FROM xx ORDER BY sin(x) DESC LIMIT 3;
  x  |        sin         
-----+--------------------
 1.4 | 0.9854497299884601
 1.3 |  0.963558185417193
   2 | 0.9092974268256817
(3 rows)
```

Given an xx table with a float x column. There is no secret exchange of information between ORDER BY and sin(x).

nyh commented on cql3/functions/vector_search_fcts.cc:
```diff
@@ -0,0 +1,61 @@
+namespace {
+
+static size_t find_matching_key_index(const expr::evaluation_inputs& inputs, const std::vector<cql3::statements::primary_key>& keys) {
+    for (size_t i = 0; i < keys.size(); ++i) {
```
I don't understand what I'm seeing here. What is the "keys" vector - how big is it? Is it the entire O(N) result list? So this function is O(N)? And we call it for every result? If so, this code has quadratic complexity, which is pretty bad.
Or am I misunderstanding what this function does? Maybe it's a sign it needs a comment.

nyh reviewed (commented):
I left a few comments, but two things worry me:
1. It seems to me that the algorithm you used makes `SELECT x, vector_similarity() ANN ...` be quadratic (O(N^2)) on the number of results. Did I understand correctly? If I understood correctly, I don't think it needs to be this way (I suggested how I think we should process the results as pairs and not two separate lists).
2. You added a few vector-search specific stuff to expressions and queries, that I don't feel very comfortable reviewing, I think @avikivity should also take a look at that part at least. I feel like these changes should not have been necessary, but I'm not really confident in this feeling.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -1670,7 +1670,7 @@ static cql3::raw_value do_evaluate(const function_call& fun_call, const evaluati
     }
 
-    bytes_opt result = scalar_fun->execute(arguments);
+    bytes_opt result = scalar_fun->execute(arguments, inputs);
```
This is very wrong. A function should be of its inputs, not of its inputs and the entire world.

Of course, you don't want to compute this function, you want to make use of the pre-computed similarity metric provided by the index.

What we might do is replace the function call by a cql3::expr::temporary, and feed the temporary into evaluation_inputs::temporaries. Alternatively (maybe better) have a fake column for the metrics, and substitute a cql3::expr::column_value for the function call. Not sure which is more difficult.

nyh replied:
"fake columns" is also more-or-less what I imagined we could have done. Imagine that someone does "select x, y, vector_distance" and for each row, x and y are real columns but the vector_distance isn't a real column - but rather something calculated during the ANN search. Each result row would have its own value of that "fake column", so even if we do filtering over the rows we will still get the correct columns with the correct distances - and don't need the O(N^2) algorithm for matching the keys.

avikivity replied:
Yes, we already have two cases of computed columns. Here it may be more difficult though. Using a temporary may be easier.

avikivity commented on db/functions/scalar_function.hh:
```diff
@@ -25,7 +26,7 @@ public:
      * @throws InvalidRequestException if this function cannot not be applied to the parameter
      */
-    virtual bytes_opt execute(std::span<const bytes_opt> parameters) = 0;
+    virtual bytes_opt execute(std::span<const bytes_opt> parameters, const cql3::expr::evaluation_inputs& inputs) = 0;
```
It's a big layering violation.

QuerthDP commented on cql3/statements/select_statement.cc:
```diff
@@ -1221,23 +1221,25 @@ indexed_table_select_statement::actually_do_execute(query_processor& qp,
         auto as = abort_source();
-        auto pkeys = co_await qp.vector_store_client().ann(_schema->ks_name(), _index.metadata().name(), _schema , std::move(ann_vector), limit, as);
-        if (!pkeys.has_value()) {
+        auto ann_result = co_await qp.vector_store_client().ann(_schema->ks_name(), _index.metadata().name(), _schema , std::move(ann_vector), limit, as);
```
I used the singular form as it's the single result (response) from ANN query.
Of course it does contain multiple rows so it feels more natural to have it in plural form.
I guess we may treat it as the list of results (found vectors) and use the plural form.

QuerthDP commented on test/boost/vector_store_client_test.cc:
```diff
@@ -466,71 +466,72 @@ SEASTAR_TEST_CASE(vector_store_client_test_ann_request) {
-                BOOST_CHECK_EQUAL(seastar::format("{}", keys->at(1).clustering.explode()), "[01, 03]");
+                ann_result = co_await vs.ann("ks", "idx", schema, std::vector<float>{0.1, 0.2, 0.3}, 2, as.as);
+                BOOST_REQUIRE(ann_result);
+                auto [keys, _] = ann_result.value();
```
You're right. It should be checked here.

QuerthDP commented on vector_search/vector_store_client.cc:
```diff
@@ -156,11 +158,19 @@ auto ck_from_json(rjson::value const& item, std::size_t idx, schema_ptr const& s
     return clustering_key_prefix::from_exploded(raw_ck);
 }
 
+auto distance_from_json(rjson::value const& item) -> std::expected<float, ann_error> {
```
I used the syntax to match the style with all other functions in this file.
It is weird to be the outstander 😅

> (regrefully I see there are many of these in vector search code, but you can see they rare elsewhere - although I was actually surprised we do have a few of them, I wonder why - is this a new fashion?)

I believe this comes from the Rust style. I think it's @ewienik's matter of habit, but maybe there is a reason for this.

QuerthDP commented on cql3/functions/vector_search_fcts.hh:
```diff
@@ -0,0 +1,35 @@
+ */
+
+/*
+ * SPDX-License-Identifier: (LicenseRef-ScyllaDB-Source-Available-1.0 and Apache-2.0)
```
What is our licensing policy?
Should all new files be marked with:
```
/*
 * Copyright (C) 2025-present ScyllaDB
 */

/*
 * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
 */
```

Or are there any restrictions?
I see that most of the functions/ files have both licences.

QuerthDP replied:
Fixed the header with correct license

QuerthDP commented:
Changelog:

- fixed licence header in new files
- renamed `ann_result` -> `ann_results`
- the distances value is now not ignored in `vector_store_client_test`
- substituted `vector_similarity()` function with `similarity_[metric]([vector_column], [vector_target_literal])`
- added arg types check for similarity functions

TODO:

- fix the problem with passing evaluation inputs (temporary column (?))
- change `ann_results` from pair of vectors to vector of pairs
- improve the O(n^2) complexity for key matching
- ensure the functions are only run with correct arguments and similarity metric matching to the vector index's
- add tests for invalid calls
- adjust integration tests as the functions did change
- add documentation
- ...

QuerthDP commented:
Changelog:

- rebased on master

QuerthDP commented:
Changelog:

- adjust CQL grammar to allow users pass multiple arguments to the similarity functions then to be informed about the syntax by an exceptions
- add more verbose exceptions for invalid function calls
- implement tests for invalid function call exceptions (working and checked on Cassandra as well)

QuerthDP commented:
Changelog:
- run `clang-format`

QuerthDP commented:
Changelog:
- fix `.rst` formatting for the Docs job to pass

QuerthDP commented:
Changelog:
- resolve Copilot's remarks

nyh commented on cql3/Cql.g:
```diff
@@ -2371,6 +2394,10 @@ K_EXECUTE:     E X E C U T E;
 
+K_SIMILARITY_EUCLIDEAN:     S I M I L A R I T Y '_' E U C L I D E A N;
+K_SIMILARITY_COSINE:        S I M I L A R I T Y '_' C O S I N E;
+K_SIMILARITY_DOT_PRODUCT:   S I M I L A R I T Y '_' D O T '_' P R O D U C T;
```
Maybe should add these new names to basic_unreserved_keyword?

QuerthDP replied:
Not really sure about this.
Neither `K_COUNT` nor `K_TOKEN` are listed there for example.

Moreover Cassandra does not even have any of these `K_SIMILARITY_*` in their Lexer.g, maybe shouldn't we as well.
But I believe the distinction between those functions and standard "column-as-argument-functions" is needed in our case, or at least much simpler and does not require serious refactoring.

avikivity replied:
It's better to move them to to unreserved keywords. Otherwise they become reserved and if you use those names at random places things stop working. It's unlikely someone named a column similarity_dot_product, but we don't want their queries to break if they did.

If it causes a grammar ambiguity,  we can keep them reserved and take the risk.

COUNT and TOKEN are very old and therefore they could have been reserved.

Anything that isn't a production in the grammar is automatically not a reserved keyword. unreserved_keywords provides an escape so the grammar can "unrecognize" a keyword to be used as an identifier.

nyh commented on cql3/statements/primary_key.hh:
```diff
@@ -16,9 +16,34 @@ namespace cql3 {
 
 /// Encapsulates a partition key and clustering key prefix as a primary key.
-struct primary_key {
+class primary_key {
```
Comment about the commit message and the approach, not this specific line:

I remember being bothered in the previous patch by the need to do a O(n) search for each key making the whole thing O(N^2) and I agree that searching with a hash table makes each search O(1) and the whole thing O(N) so it's much better. But I still don't understand why we need to search for the keys at all - why we don't just keep the  seartch results with the scores (each match score is like an extra column) and we wouldn't need those lookups all.

The need for this "key matching" also forced you to add unselected key columns to the select (in the next patch), and could have been avoiding if "key matching" wasn't needed.

QuerthDP replied:
> But I still don't understand why we need to search for the keys at all - why we don't just keep the seartch results with the scores (each match score is like an extra column) and we wouldn't need those lookups all.

I don't really understand your idea here. Would be really great if you could explain it a bit more.
As far as I understand, in `do_execute` of `select_statement` and other related stuff, we only provide a primary keys (or partition key ranges) to inform the later execution which rows should be returned in the result.
I don't see a place there that we could store/pass/provide the distances with keys, and I believe passing the pair as a key and some additional value is also not an option (unnecessary in probably all cases apart from this one).
I feel like your idea leads to having a fake column as you mentioned before in this conversation, but I quite don't understand the implementation details here yet. Should probably take more time to research on that, but your help would be really appreciated if you have some solution in your head more crystalized.

I though that it would be possible to defer this by improving the complexity (from O(n^2) to O(n) is a great result, but low hanging fruit in this case), but I believe there may be some classical rules why having it done the way I did should be avoided.

nyh commented on test/cqlpy/test_vector_index.py:
```diff
@@ -337,3 +337,141 @@ def test_vector_search_when_tracing_is_enabled(cql, test_keyspace, scylla_only):
+###############################################################################
+
+
+# The test is Scylla-only because Cassandra does not require vector similarity functions to be used only with ANN queries.
```
So why do we require that?

QuerthDP replied:
We do not calculate the results in Scylla, they are acquired from the ANN results returned by Vector Search service.
No ANN query = No results
Cassandra AFAIK does calculate the results (or have it precomputed somewhere in the SAI), so that they may return any similarity function of any vectors in any query - we have limitation on that, as the Vector Search service store only one index with a specific similarity metric set.

Nevertheless we could allow for such calculations (and might do in the future), but at the actual state it would have a massive impact on performance, so we'd rather disallow this.

nyh commented on test/cqlpy/test_vector_index.py:
```diff
@@ -337,3 +337,141 @@ def test_vector_search_when_tracing_is_enabled(cql, test_keyspace, scylla_only):
+        cql.execute(f"DROP INDEX {test_keyspace}.{table.split('.')[1]}_v_idx")
+
+
+# The test is Scylla-only because Cassandra does not require the arguments to be the same as for the ANN query.
```
Do you know why Cassandra doesn't require this?
Again it's not clear to me if "scylla_only" is the right verdict on this test - or if it should be xfail (a Scylla bug) or cassandra_bug (if it's a Cassandra bug), or what...

QuerthDP commented on test/vector_search/vector_store_client_test.cc:
```diff
@@ -953,3 +961,149 @@ SEASTAR_TEST_CASE(vector_search_metrics_test) {
+                                "SELECT pk1, similarity_{}(v2, [0.1, 0.2, 0.3]) FROM ks.test ORDER BY v1 ANN OF [0.1, 0.2, 0.3] LIMIT 1", similarity_function));
+                        BOOST_FAIL("Expected invalid_request_exception was not thrown");
+                    } catch (exceptions::invalid_request_exception& ex) {
+                        BOOST_CHECK_MESSAGE(
```
Thank you, that looks much better.
Apart from that - great catch! The `BOOST_CHECK_MESSAGE` would not fail if the message was different.
The `BOOST_CHECK_EQUAL` is the correct approach.

QuerthDP commented on test/vector_search/vector_store_client_test.cc:
```diff
@@ -953,3 +961,149 @@ SEASTAR_TEST_CASE(vector_search_metrics_test) {
+                             fmt::format("CREATE CUSTOM INDEX idx2 ON ks.test (v2) USING 'vector_index' WITH OPTIONS = {{'similarity_function': '{}'}}",
+                                     similarity_function)))
+                            .get()
+                            ->throw_if_exception();
```
Wanted to have the exception thrown on failure, but it looks like without the `get()->throw_if_exception()` it also works as intended.

QuerthDP commented on docs/cql/functions.rst:
```diff
@@ -227,6 +227,39 @@ A number of functions are provided to “convert” the native types into binary
+    The ``vector_column`` and the ``vector_value`` must be **exactly the same** as used in the ``ORDER BY ... ANN OF ...`` clause of the query.
+    Otherwise the query will fail with an error.
+
+The functions return a ``float`` value representing the similarity distance fetched from Vector Search service while executing the ANN vector query.
```
Removed the mention of Vector Search service. Added some more information about distances

QuerthDP commented:
Changelog:
- fixed `vector_search/` similarity function tests which mistakenly used `BOOST_CHECK_MESSAGE`
- refactored the  `vector_search/` similarity function tests according to @knowack1's remarks
- added a simple check for good ANN query with Cassandra
- refactored the `vector_search_fct` creation in `functions.cc`
- adjusted docs to explain "distance" a bit more

avikivity commented on cql3/Cql.g:
```diff
@@ -451,13 +451,14 @@ selector returns [shared_ptr<raw_selector> s]
+       | K_TTL       '(' c=cident ')'                                  { tmp = column_mutation_attribute{column_mutation_attribute::attribute_kind::ttl,
+                                                                                                         unresolved_identifier{std::move(c)}}; }
+       | f=functionName           args=selectionFunctionArgs           { tmp = function_call{std::move(f), std::move(args)}; }
+       | f=similarityFunctionName args=vectorSimilarityArgs            { tmp = function_call{std::move(f), std::move(args)}; }
```
The problem is the vector argument per se, it's the fact that selectionFunctionArgs don't accept a `constant` or a `marker`.

I won't require to fix that brokenness in the grammar now, but at least leave a comment explaining the reason for the duplication.

avikivity commented on cql3/statements/primary_key.hh:
```diff
@@ -16,9 +16,34 @@ namespace cql3 {
 
 /// Encapsulates a partition key and clustering key prefix as a primary key.
-struct primary_key {
+class primary_key {
```
It's still a struct in spirit. public members -> struct, internals != interface -> class.

QuerthDP replied:
Refactored the commit, now it does not mention the change.

QuerthDP commented on cql3/statements/primary_key.hh:
```diff
@@ -16,9 +16,34 @@ namespace cql3 {
     clustering_key_prefix clustering;
+
+    primary_key(const dht::decorated_key& partition, const clustering_key_prefix& clustering)
+        : partition(partition), clustering(clustering) {}
```
You're right. Changed to struct, class is not needed in this case.

nyh commented on cql3/query_options.cc:
```diff
@@ -118,6 +118,18 @@ query_options::query_options(std::unique_ptr<query_options> qo, lw_shared_ptr<se
 
 }
 
+query_options::query_options(std::unique_ptr<query_options> qo, statements::raw::select_statement::prepared_ann_ordering_type ann_ordering, vector_search::vector_store_client::ann_results ann_results)
```
I think Avi's suggestion is more-or-less what I proposed above, except I didn't think that actually a new **schema** would be necessary. Already today it is possible for SELECT to return things which aren't in the schema, such as functions (e.g., SELECT TTL(x)) and other stuff, I thought it shouldn't be hard to add to the output of the select also that similarity column without making it part of any scehema.
